### PR TITLE
DMF User-mode Test Program and other Support

### PR DESCRIPTION
**Pull request by samtertzakian**
1. Add DmfUTest driver to unit test Modules in User-mode.
2. Update DMF_Registry Module to support User-mode.
3. Correct issue in random number generating function in Test driver.
4. Clean up DmfKTest.inf file.

### CODE REVIEW

File: Dmf/Framework/Modules.Core/Dmf_BufferPool.c [hunk 1]
```diff
@@ -1107,6 +1107,20 @@ Return Value:
               (moduleConfig->BufferPoolMode == BufferPool_Mode_Sink && moduleConfig->Mode.SourceSettings.BufferCount == 0));
     moduleContext->NumberOfBuffersSpecifiedByClient = moduleConfig->Mode.SourceSettings.BufferCount;
 
+#if defined(DMF_USER_MODE)
+    // It is not possible to use "PutWithTimer" Method when lookaside list is enabled in User-mode
+    // because buffers are deleted in the timer callback which causes the child WDFTIMER to also
+    // be deleted. That, in turn, can cause a deadlock and verifier issue.
+    //
```
Comment (ankurung) [inline Dmf/Framework/Modules.Core/Dmf_BufferPool.c]: Do we want to update the md files as well?

File: Dmf/Modules.Library.Tests/Dmf_Tests_BufferPool.c [hunk 2]
```diff
@@ -766,7 +774,13 @@ Return Value:
     moduleConfigBufferPool.Mode.SourceSettings.BufferSize = BUFFER_SIZE;
     moduleConfigBufferPool.Mode.SourceSettings.BufferCount = BUFFER_COUNT_PREALLOCATED;
     moduleConfigBufferPool.Mode.SourceSettings.CreateWithTimer = TRUE;
+#if !defined(DMF_USER_MODE)
+    // TODO: Cannot user "lookaside" in User-mode with timer since deleting buffers
```
Comment (ankurung) [inline Dmf/Modules.Library.Tests/Dmf_Tests_BufferPool.c]: use

File: Dmf/Modules.Library.Tests/Dmf_Tests_Registry.c [hunk 3]
```diff
@@ -1763,99 +1796,153 @@ Return Value:
...
         // Note: PLUGPLAY_REGKEY_CURRENT_HWPROFILE may not be used alone.
         //
         PLUGPLAY_REGKEY_DEVICE | PLUGPLAY_REGKEY_CURRENT_HWPROFILE,
-        PLUGPLAY_REGKEY_DRIVER | PLUGPLAY_REGKEY_CURRENT_HWPROFILE
+        PLUGPLAY_REGKEY_DRIVER | PLUGPLAY_REGKEY_CURRENT_HWPROFILE,
+#else
+        // Note: PLUGPLAY_REGKEY_CURRENT_HWPROFILE may not be used alone.
+        //
+        PLUGPLAY_REGKEY_DEVICE | WDF_REGKEY_DEVICE_SUBKEY,
+        PLUGPLAY_REGKEY_DRIVER | WDF_REGKEY_DEVICE_SUBKEY
```
Comment (ankurung) [inline Dmf/Modules.Library.Tests/Dmf_Tests_Registry.c]: This should be WDF_REGKEY_DRIVER_SUBKEY

File: Dmf/Modules.Library.Tests/Dmf_Tests_BufferPool.c (hunk 2 above)
Comment (samtertzakian) [reply Dmf/Modules.Library.Tests/Dmf_Tests_BufferPool.c]: fixed

File: Dmf/Modules.Library.Tests/Dmf_Tests_Registry.c (hunk 3 above)
Comment (samtertzakian) [reply Dmf/Modules.Library.Tests/Dmf_Tests_Registry.c]: fixed

File: Dmf/Modules.Library.Tests/TestsUtility.c [hunk 4]
```diff
@@ -61,28 +61,49 @@ Routine Description:
...
+        DmfAssert(FALSE);
+        random = 0;
+    }
+#endif
 
     return random;
 }
 
+#if defined(DMF_USER_MODE)
+#define BYTE_MAX    ((ULONG)0xFFFFFFFF)
```
Comment (samtertzakian) [inline Dmf/Modules.Library.Tests/TestsUtility.c]: This should be 0xFF.

**Pull request merged**